cps/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadDump.java

```diff
@@ -0,0 +1,139 @@
+package org.jenkinsci.plugins.workflow.cps;
+
+import org.jenkinsci.plugins.workflow.steps.StepExecution;
+
+import java.io.PrintStream;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+
+/**
+ * Immutable snapshot of a state of {@link CpsThreadGroup}.
+ *
+ * @author Kohsuke Kawaguchi
+ */
+public final class CpsThreadDump {
+    private final List<ThreadInfo> threads = new ArrayList<ThreadInfo>();
+
+    public static final class ThreadInfo {
+        private final String headline;
+        private final List<StackTraceElement> stack = new ArrayList<StackTraceElement>();
+
+        /**
+         * Given a list of {@link CpsThread}s that share the same {@link FlowHead}, in the order
+         * from outer to inner, reconstruct the thread stack.
+         */
+        private ThreadInfo(List<CpsThread> e) {
+            CpsThread head = e.get(e.size() - 1);
+            headline = head.toString();
+
+            ListIterator<CpsThread> itr = e.listIterator(e.size());
+            while (itr.hasPrevious()) {
+                CpsThread t = itr.previous();
+
+                StepExecution s = t.getStep();
+                if (s !=null) {
+                    stack.add(new StackTraceElement(s.getClass().getName(), "<>", null, -2));
```

Comment:
I think the `functionName` would be more appropriate here. People do not generally care about the implementation class.